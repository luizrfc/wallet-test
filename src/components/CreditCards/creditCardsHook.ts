import { Colors } from "../../styles/themeStyle";

const getNameCard = (_kind: string) => {
  if (_kind === "black") return "Black Card";
  if (_kind === "green") return "Green Card";
  return "Credit Card";
};

const getColorCard = (_kind: string) => {
  if (_kind === "black") return Colors.black;
  if (_kind === "green") return Colors.primary;
  return Colors.primary;
};

const getColorFontCard = (_kind: string) => {
  if (_kind === "black") return Colors.white;
  return Colors.black;
};

export { getColorCard, getColorFontCard, getNameCard };
