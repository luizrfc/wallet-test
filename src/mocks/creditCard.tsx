import { ICardModel } from '../models/cardModel';

export const mockCreditCard: ICardModel = {
    number: "5502 0919 4514 2207",
    cvv: "576",
    name: "John Doe",
    expiry: "12/25",
    kind: "green"
}