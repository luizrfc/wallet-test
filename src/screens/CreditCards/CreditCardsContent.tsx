import React, { Fragment, useContext } from "react";

import { Button, Container, CreditCards, Text } from "../../components";

import { TouchableOpacity, View } from "react-native";
import Animated, { FadeIn, FadeInUp } from "react-native-reanimated";
import { Colors, styleScreen } from "../../styles/themeStyle";
import useCreditCards from "./CreditCardsHook";
import { CreditCardContext } from "../../context/creditCardsContext";

function CreditCardsContent() {
    const { cardSelected, animatedCardOne, animatedCardTwo, selectCard } = useCreditCards();
    const creditCardContext = useContext(CreditCardContext);

    return (
        <Container ph={0}>
            <Fragment>
                <View
                    style={{
                        width: "100%",
                        backgroundColor: Colors.white,
                        paddingVertical: 24,
                        borderBottomRightRadius: 48,
                        borderBottomLeftRadius: 48,
                        justifyContent: "center",
                        alignItems: "center"
                    }}
                >
                    <Text color={Colors.secondary} text="Meus cartões" type="h4" />
                </View>
                <View
                    style={[
                        styleScreen.screen,
                        { paddingVertical: 16, paddingHorizontal: 32 }
                    ]}
                >
                    {creditCardContext?.data.map((card) => (
                        <Animated.View
                            style={[
                                { width: "100%", position: "absolute" },
                                card.position == 1
                                    ? animatedCardOne
                                    : animatedCardTwo
                            ]}
                            key={card.position}
                        >
                            <TouchableOpacity
                                activeOpacity={0.9}
                                onPress={() => selectCard(card.position)}
                            >
                                <CreditCards
                                    name={card.name}
                                    number={card.number}
                                    expiry={card.expiry}
                                    id={card.id}
                                    kind={card.kind}
                                />
                            </TouchableOpacity>
                        </Animated.View>
                    ))}

                    {cardSelected > 2 ? (
                        <Animated.View
                            entering={FadeInUp}
                            style={{
                                width: "100%",
                                position: "absolute",
                                bottom: 200,
                                alignItems: "center"
                            }}
                        >
                            <Button text="pagar com este cartão" type="secondary" />
                        </Animated.View>
                    ) : (
                        <Animated.View
                            entering={FadeIn}
                            style={{
                                width: "100%",
                                position: "absolute",
                                bottom: 200,
                                alignItems: "center"
                            }}
                        >
                            <Text
                                text="usar este cartão"
                                color={Colors.white}
                                type="paragraph"
                            />
                        </Animated.View>
                    )}
                </View>
            </Fragment>
        </Container>
    );
}

export default CreditCardsContent;
