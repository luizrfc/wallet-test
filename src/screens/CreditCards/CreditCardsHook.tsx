import { ReactElement, useState } from "react";
import { Dimensions } from "react-native";
import {
  Extrapolate,
  interpolate,
  useAnimatedStyle,
  withSpring
} from "react-native-reanimated";
import { ICardModel } from "../../models/cardModel";
import { getCards } from "../../services/cardsService";

interface NavigationProps {
  children: ReactElement<any, any>;
}

interface ICards extends ICardModel {
  position: number;
}

interface ICreditCardsHooks {
  data: Array<ICards>;
  cardSelected: number;
  selectCard: (value: number) => void;
  startScreen: () => void;
  animatedCardOne: any;
  animatedCardTwo: any;
}

const useCreditCards = (): ICreditCardsHooks => {
  const [data, setData] = useState<Array<ICards>>([]);
  const [cardSelected, setCardSelected] = useState<number>(1);

  const getCreditCards = async () => {
    const result = await getCards();
    if (result.success) {
      const cardsList = result.data.map((card: any, index: number) => ({
        ...card,
        position: index + 1
      }));
      setData(cardsList);
    }
  };

  const configSpring = {
    duration: 1000,
    stiffness: 100,
    overshootClamping: false,
    restDisplacementThreshold: 0.01,
    restSpeedThreshold: 25
  };

  const deviceHeight = Dimensions.get("window").height - 250;

  const animatedCardOne = useAnimatedStyle(() => ({
    top: withSpring(
      interpolate(
        cardSelected ?? 1,
        [1, 2, 3, 4],
        [180, 120, 100, deviceHeight],
        Extrapolate.CLAMP
      ),
      configSpring
    ),
    zIndex: withSpring(
      interpolate(cardSelected ?? 1, [1, 2], [1, 0], Extrapolate.CLAMP)
    ),
    opacity: withSpring(
      interpolate(
        cardSelected ?? 1,
        [1, 2, 3, 4],
        [1, 1, 1, 0.6],
        Extrapolate.CLAMP
      )
    )
  }));

  const animatedCardTwo = useAnimatedStyle(() => ({
    top: withSpring(
      interpolate(
        cardSelected ?? 2,
        [1, 2, 3, 4],
        [120, 180, deviceHeight, 100],
        Extrapolate.CLAMP
      ),
      configSpring
    ),
    zIndex: withSpring(
      interpolate(cardSelected ?? 2, [1, 2], [0, 1], Extrapolate.CLAMP)
    ),
    opacity: withSpring(
      interpolate(
        cardSelected ?? 1,
        [1, 2, 3, 4],
        [1, 1, 0.6, 1],
        Extrapolate.CLAMP
      )
    )
  }));

  const selectCard = (position: number) => {
    if (position === 1) {
      setCardSelected(cardSelected === 4 ? 1 : 3);
    }
    if (position === 2) {
      setCardSelected(cardSelected === 3 ? 2 : 4);
    }
  };

  const startScreen = () => {
    setCardSelected(1);
    getCreditCards();
  };

  return {
    data,
    cardSelected,
    animatedCardOne,
    animatedCardTwo,
    selectCard,
    startScreen
  }
};

export default useCreditCards;
