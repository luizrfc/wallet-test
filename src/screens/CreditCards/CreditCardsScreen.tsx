import { View } from "react-native";
import { Box, Button, Container, Text } from "../../components";
import { NavigationProp, useNavigation } from "@react-navigation/native";
import { RootStackParamList } from "../../navigation/navigationStackParams";

const CreditCardsScreen = (): JSX.Element => {
    const navigation = useNavigation<NavigationProp<RootStackParamList>>();

    const handleClick = () => {
        navigation.navigate("Home");
    };

    return <Container>
        <Text text="CartÃµes" type="h1" weight="regular" />
        <Box>
            <Button
                text="Home"
                type="secondary"
                handleClick={() => handleClick()}
            />
        </Box>
    </Container>
}

export default CreditCardsScreen;