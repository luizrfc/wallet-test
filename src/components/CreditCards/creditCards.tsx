import React from "react";
import { View } from "react-native";
import { ICardModel } from "../../models/cardModel";
import Text from "../TextComponent";
import Item from "../Containers/item";
import { anonymizeNumberCard } from "../../utils/validate";
import { getColorCard, getColorFontCard, getNameCard } from "./creditCardsHook";


const CreditCards = ({
  name = "Customer",
  number = "0000 0000 0000 0000",
  expiry = "00/00",
  kind = "black"
}: ICardModel): React.ReactNode => {

  return (
    <View
      style={{
        width: "100%",
        paddingHorizontal: 16,
        paddingVertical: 30,
        marginVertical: 20,
        borderRadius: 16,
        backgroundColor: getColorCard(kind)
      }}
    >
      <Text text={getNameCard(kind)} color={getColorFontCard(kind)} type="h5" />
      <Item mt={34} align="left">
        <Text text={name} color={getColorFontCard(kind)} type="paragraph" />
      </Item>
      <Item mt={4} align="left">
        <Text
          text={anonymizeNumberCard(number)}
          color={getColorFontCard(kind)}
          type="small"
        />
      </Item>
      <Item mt={4} align="left">
        <Text
          text={`Validade ${expiry}`}
          color={getColorFontCard(kind)}
          type="small"
        />
      </Item>
    </View>
  );
};

export default CreditCards;
