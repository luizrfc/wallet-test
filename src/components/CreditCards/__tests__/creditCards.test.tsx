import React from "react";
import "react-native";
import renderer from "react-test-renderer";
import CreditCards from "../creditCards";
import { mockCreditCard } from "../../../mocks/creditCard";

const mockCard = mockCreditCard;

test("Render a credit card correctly", () => {
  const button = renderer
    .create(
      <CreditCards
        name={mockCard.name}
        number={mockCard.number}
        expiry={mockCard.expiry}
        id={mockCard.id}
        kind={mockCard.kind}
      />
    )
    .toJSON();
  expect(button).toMatchSnapshot();
});
