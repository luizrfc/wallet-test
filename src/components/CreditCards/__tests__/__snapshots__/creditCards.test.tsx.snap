// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render a credit card correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#A5FF32",
      "borderRadius": 16,
      "marginVertical": 20,
      "paddingHorizontal": 16,
      "paddingVertical": 30,
      "width": "100%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000",
          "fontFamily": "PTSans-CaptionRegular",
        },
        {
          "fontSize": 18,
          "lineHeight": 20,
        },
      ]
    }
  >
    Green Card
  </Text>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        {
          "marginBottom": 6,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 34,
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "PTSans-CaptionRegular",
          },
          {
            "fontSize": 16,
            "lineHeight": 18,
          },
        ]
      }
    >
      John Doe
    </Text>
  </View>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        {
          "marginBottom": 6,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "PTSans-CaptionRegular",
          },
          {
            "fontSize": 14,
            "lineHeight": 16,
          },
        ]
      }
    >
      •••• •••• •••• 2207
    </Text>
  </View>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        {
          "marginBottom": 6,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "PTSans-CaptionRegular",
          },
          {
            "fontSize": 14,
            "lineHeight": 16,
          },
        ]
      }
    >
      Validade 12/23
    </Text>
  </View>
</View>
`;

exports[`Render button correctly 1`] = `
<View
  style={
    {
      "backgroundColor": "#A5FF32",
      "borderRadius": 16,
      "marginVertical": 20,
      "paddingHorizontal": 16,
      "paddingVertical": 30,
      "width": "100%",
    }
  }
>
  <Text
    style={
      [
        {
          "color": "#000",
          "fontFamily": "PTSans-CaptionRegular",
        },
        {
          "fontSize": 18,
          "lineHeight": 20,
        },
      ]
    }
  >
    Green Card
  </Text>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        {
          "marginBottom": 6,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 34,
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "PTSans-CaptionRegular",
          },
          {
            "fontSize": 16,
            "lineHeight": 18,
          },
        ]
      }
    >
      John Doe
    </Text>
  </View>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        {
          "marginBottom": 6,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "PTSans-CaptionRegular",
          },
          {
            "fontSize": 14,
            "lineHeight": 16,
          },
        ]
      }
    >
      •••• •••• •••• 2207
    </Text>
  </View>
  <View
    style={
      [
        {
          "width": "100%",
        },
        {
          "paddingBottom": 0,
          "paddingLeft": 0,
          "paddingRight": 0,
          "paddingTop": 0,
        },
        {
          "marginBottom": 6,
          "marginLeft": 0,
          "marginRight": 0,
          "marginTop": 4,
        },
        {
          "alignItems": "flex-start",
        },
      ]
    }
  >
    <Text
      style={
        [
          {
            "color": "#000",
            "fontFamily": "PTSans-CaptionRegular",
          },
          {
            "fontSize": 14,
            "lineHeight": 16,
          },
        ]
      }
    >
      Validade 12/23
    </Text>
  </View>
</View>
`;
